### PR TITLE
mgmt: mcumgr: Deprecate returning rc responses when status is OK.

### LINKED ISSUE #42590: mgmt/mcumgr/lib: RFC: Allow leaving out "rc" in successful respones and use "rc" only for SMP processing errors.
**Is your enhancement proposal related to a problem? Please describe.**
There is inconsistency on how mcumgr responds to commands regarding usage of "rc" field.
When there are simple error responses, it is quite consistent and error responses usually consist only of map containing "rc" as a key and error code (int) as value.
Successful responses are not that consistent for example:
 - reset command returns empty dictionary on success;
 - echo does not return "rc" on success, only echo string;
 - img update command will include "rc" in every response, but image list only in case of error; image erase always return "rc";
 - stat commands always include "rc" = 0 in successful response, for all commands;

Additional problem is that some commands may use non-0 "rc" in successful responses; shell is an example when successful response, where SMP request has been properly decoded executed and response encoded, can have "rc" error because even though SMP processing went fine the executed command returned error that is returned as "rc" - which is not correct usage!

**Describe the solution you'd like**
Consistency: allow to leave out "rc" if it is 0 (success), and if "rc" is only response allow to send no payload at all, header should be enough. Only SMP processing errors should trigger addition of "rc" field to response.

**Describe alternatives you've considered**
Leaving current state.

**Additional context**
TBD

### PR DESCRIPTION
**Pull request by nordicjm**
This change deprecates the previous mcumgr behaviour of return result codes when the status is 0 (OK), instead it will skip the rc field for these responses. If there is only an rc field with status 0 to return, then mcumgr will now instead just return an empty map.

Fixes #42590

- [X] Remove rc responses when status is 0
- [X] Add deprecation warning/notice
- [X] Release notes update
- [X] Documentation update
- [X] Manual tests (mcumgr command line tool functions fine with new behaviour)

### CODE REVIEW

**Review by de-nordic — CHANGES_REQUESTED** [review]

File: subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt.c [hunk 1]
```diff
@@ -303,23 +303,33 @@ img_mgmt_erase(struct smp_streamer *ctxt)
...
-	     zcbor_size_put(zse, g_img_mgmt_state.off);
+	if (IS_ENABLED(CONFIG_MCUMGR_CMD_LEGACY_RC_BEHAVIOUR)) {
+		ok = zcbor_tstr_put_lit(zse, "rc")		&&
+		     zcbor_int32_put(zse, MGMT_ERR_EOK);
+	}
+
+	if (ok == true) {
+		ok = zcbor_tstr_put_lit(zse, "off")		&&
+		     zcbor_size_put(zse, g_img_mgmt_state.off);
+	}
```
[the remark this answers is not in the document]
Comment (nordicjm) [reply subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt.c:332]: This reverses the order of parameters, have simplified it but kept the order the same

**Review by de-nordic — DISMISSED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by carlescufi — CHANGES_REQUESTED** [review]
No need to deprecate, this is unstable. Please remove it altogether so we can declare it stable soon.

**Comment by de-nordic** [comment]
> No need to deprecate, this is unstable. Please remove it altogether so we can declare it stable soon.

Yes, but @nordicjm mentions important issue of no-Zephyr based tools, that may require some time to adapt.
I agree with @nordicjm that we should use deprecation process as a way for external projects to adapt.

File: subsys/mgmt/mcumgr/Kconfig [hunk 2]
```diff
@@ -125,6 +125,19 @@ config MCUMGR_SMP_COMMAND_STATUS_HOOKS
 	  This will enable SMP command status notification hooks for when an SMP message is
 	  received or processed.
 
+config MCUMGR_SMP_LEGACY_RC_BEHAVIOUR
+	bool "Legacy rc (result code) response behaviour (DEPRECATED)"
+	select DEPRECATED
```
Comment (carlescufi) [inline subsys/mgmt/mcumgr/Kconfig:130]: Please remove this `select` statement.

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: subsys/mgmt/mcumgr/Kconfig (hunk 2 above)
Comment (nordicjm) [reply subsys/mgmt/mcumgr/Kconfig]: Fixed

**Review by de-nordic — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**